Imx296

naushir opened the pull request:
This PR removes the downstream imx296 driver and replaces it with the upstream version.  Some additional features/fixes have also been added:

- H/V flip support.
- Sensor crop support.
- Fix gain delay to 1 frame.
- Disable binning on colour sensor variant.

Mani-Sadhasivam committed:
dt-bindings: media: i2c: Add IMX296 CMOS sensor binding

Add YAML devicetree binding for IMX296 CMOS image sensor. Let's also
add MAINTAINERS entry for the binding and driver.

Signed-off-by: Manivannan Sadhasivam <manivannan.sadhasivam@linaro.org>
Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>
Reviewed-by: Rob Herring <robh@kernel.org>
Signed-off-by: Sakari Ailus <sakari.ailus@linux.intel.com>

pinchartl committed:
media: i2c: IMX296 camera sensor driver

The IMX296LLR is a monochrome 1.60MP CMOS sensor from Sony. The driver
supports cropping and binning (but not both at the same time due to
hardware limitations) and exposure, gain, vertical blanking and test
pattern controls.

Preliminary support is also included for the color IMX296LQR sensor.

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>
[Sakari Ailus: Make driver's remove function return void]
Signed-off-by: Sakari Ailus <sakari.ailus@linux.intel.com>

naushir committed:
Revert "media: i2c: imx296: Support for the Sony IMX296 sensor"

This reverts commit 26c9a943e4a9fde3355dd12a3127f2f22f7d5551.

naushir committed:
media: i2c: imx296: Get sensor crop working

Add a missing register write (MIPIC_AREA3W) when setting up a crop
window in the sensor to get this functionality working.

Signed-off-by: Naushir Patuck <naush@raspberrypi.com>

naushir committed:
media: i2c: imx296: Set a 1 frame gain delay

Set the gain delay to 1 frame in the sensor. This avoids any race
condition or ambiguity over when the setting is applied through
userland.

Signed-off-by: Naushir Patuck <naush@raspberrypi.com>

naushir committed:
media: i2c: imx296: Add horizontal/vertical flip support

Add support for setting horizontal and/or vertial flips in the IMX296
sensor through the V4L2_CID_HFLIP and V4L2_CID_VFLIP controls.

Add a new helper function to return the media bus format code that
depends on the sensor flips.

Grab the V4L2_CID_HFLIP and V4L2_CID_VFLIP controls on stream on, and
release on stream off to ensure flips cannot be changed while the sensor
is streaming.

Signed-off-by: Naushir Patuck <naush@raspberrypi.com>

naushir committed:
media: i2c: imx296: Disable binning for colour variant

Binning is only supported on the mono variant of the IMX296 sensor.
Disable reporting of the binning mode for the colour variant.

Signed-off-by: Naushir Patuck <naush@raspberrypi.com>

naushir committed:
media: i2c: imx296: Add helper for hblank control

Add a helper function to setup the horizontal blanking control. Update
the control limits on set_format as the horizontal blanking time must
remain constant regardless of sensor output width.

Signed-off-by: Naushir Patuck <naush@raspberrypi.com>

6by9 commented:
I'm assuming the two base commits are 
https://github.com/torvalds/linux/commit/d680dc5805745e68e3c8d769f40685707ca7fe4f and 
https://github.com/torvalds/linux/commit/cb33db2b6ccfe3ccc13347755ab3ef38691d59c3
which are both in 6.3, but they would seem to be the patches direct from the list as they don't have Mauro's SoB.
Normal approach would be to cherry-pick and add "Upstream commit <hash>" to the commit text.

I'm not reviewing those as they should be good from upstream. Review on the other minor mods to come....

naushir commented:
I actually pulled them from linux-media/master, I assume they are the same.  They seem to match the commits in rpi-6.3.y.

6by9 commented on drivers/media/i2c/imx296.c:
```diff
@@ -384,10 +384,35 @@ static const struct v4l2_ctrl_ops imx296_ctrl_ops = {
+	/*
+	 * Horizontal blanking is controlled through the HMAX register, which
+	 * contains a line length in INCK clock units. The INCK frequency is
+	 * fixed to 74.25 MHz. The HMAX value is currently fixed to 1100,
```
I realise this is a move of code from the base driver, however INCK is not fixed at 74.25MHz. ```imx296_clk_params``` has entries for 37.125, 54, and 74.25MHz.

Use sensor->clk_params->freq instead of a hardcoded 74250000.
If I read "Readout drive modes" on page 44 of the datasheet, it lists the values for "Number of INCK in 1H" for each INCK. 
Seeing as I believe (from the overlay) we're using a 54MHz INCK, I'm slightly surprised that HMAX is correctly set when fixed at 1100.

naushir replied:
I believe this is correct.  The sensor internally multiplies any INCLK value to 74.25 MHz.  This is done by setting the INCLK value in one of the clk related registers. This can be verified by the frame time computations, which do give the right numbers, implying HMAX is at the right value.

6by9 replied:
That would imply that the comment is misleading. Not your comment, so not a blocker for this PR though.
These patches really ought to be upstreamed though, hence me being pedantic.

naushir replied:
I could reword the comment as:

```
	 * Horizontal blanking is controlled through the HMAX register, which
	 * contains a line length in contains a line length in units of an
	 * internal 74.25 MHz clock derived from the INCLK. The HMAX value is
	 * currently fixed to 1100, convert it to a number of pixels based on
	 * the nominal pixel rate.
```

naushir replied:
> These patches really ought to be upstreamed though, hence me being pedantic.

Yes, no reason not to send these patches upstream.

6by9 commented on drivers/media/i2c/imx296.c:
```diff
@@ -597,6 +646,10 @@ static int imx296_stream_on(struct imx296 *sensor)
 	imx296_write(sensor, IMX296_CTRL0A, 0, &ret);
 
+	/* vflip and hflip cannot change during streaming */
+	__v4l2_ctrl_grab(sensor->vflip, 1);
```
Does changing flip create a corrupt frame? If not then we don't need to grab the control on a mono variant of the sensor.

naushir replied:
Nothing is mentioned in the datasheet.  But I would imagine there may be some timing related issues doing it mid-frame.  To be safe, I would consider keeping this as-is.

6by9 replied:
I only raised it as I had the same question on [imx290](https://patchwork.linuxtv.org/project/linux-media/patch/20230131192016.3476937-12-dave.stevenson@raspberrypi.com/#144355) where it is explicitly stated in the datasheet that it will create a corrupt frame, however that is a rolling shutter sensor compared to global.

6by9 commented on drivers/media/i2c/imx296.c:
```diff
@@ -427,6 +466,16 @@ static int imx296_ctrls_init(struct imx296 *sensor)
 
+	sensor->hflip = v4l2_ctrl_new_std(&sensor->ctrls, &imx296_ctrl_ops,
+					  V4L2_CID_HFLIP, 0, 1, 1, 0);
+	if (sensor->hflip)
```
```if (sensor->hflip && !sensor->mono```
The mono sensor doesn't change the format.

naushir replied:
Ack.

6by9 commented on drivers/media/i2c/imx296.c:
```diff
@@ -427,6 +466,16 @@ static int imx296_ctrls_init(struct imx296 *sensor)
+	sensor->vflip = v4l2_ctrl_new_std(&sensor->ctrls, &imx296_ctrl_ops,
+					  V4L2_CID_VFLIP, 0, 1, 1, 0);
+	if (sensor->vflip)
+		sensor->vflip->flags |= V4L2_CTRL_FLAG_MODIFY_LAYOUT;
```
Ditto

naushir replied:
Ack.

naushir committed:
media: i2c: imx296: Adjust cropping limits

Through emperical testing, the sensor can crop upto a 96x88 window to
produce a valid Bayer frame. Adjust the ROIWH1_MIN ROIWV1_MIN
appropriately for this limit.

Signed-off-by: Naushir Patuck <naush@raspberrypi.com>

naushir commented:
Added one more commit on top to limit the crop window to 96x88.  Below this and the sensor does not produce valid frames.

6by9 commented:
Looks good to me.

I guess it's hard to tell if the sensor just doesn't stream, or if the interrupt rate is just too high for Unicam. It makes little difference in reality.

naushir commented:
It streams, only returns a frame with all 0 pixels :-)

pelwell commented:
You just need to rebase against the current rpi-6.1.y and I'll merge.
